### PR TITLE
Implement `Into<BigUint>` and `From<BigUint>` for `BigInteger` and `PrimeField`

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

This PR partially handles the challenges of converting values between different fields, by allowing BigInteger of different sizes to convert between each other, through an intermediate representation `BigUint`. This can also make it easy to print a field element.

More support for field conversions is possible: We could implement the same thing for `PrimeField`, which likely is a simple wrapper of this. 

Note that it is also possible to enable `From<T: To<Uint>> for U: From<Uint>` so this can make the conversion simplest. I feel it is too risky since the enforcement of which field can be applied to which object can prevent certain misuse, and this more aggressive form may sacrifice such enforcement, especially conversions between different fields are, though useful, not occurring in the majority of the code.

closes: #271 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by weikengchen** [comment]
@ValarDragon Hi Dev, there is a markdown linter error that seems strange---I feel superlinter's current docker may have a possible conflict with Rust at this moment.

File: ff/src/biginteger/macros.rs [hunk 1]
```diff
@@ -352,5 +352,48 @@ macro_rules! bigint_impl {
...
+                let mut limbs = [0u64; $num_limbs];
+
+                for i in 0..$num_limbs {
+                    tmp >>= 64;
+                    tmp <<= 64;
+
+                    limbs[i] = (tmp2 - tmp.clone()).to_u64().unwrap();
+
+                    tmp >>= 64;
+                    tmp2 = tmp.clone();
```
Comment (daira) [inline ff/src/biginteger/macros.rs:374]: I see how this works, but it seems to have quadratic cost in the number of limbs. Is there no better way?

File: ff/src/biginteger/macros.rs [hunk 2]
```diff
@@ -352,5 +352,48 @@ macro_rules! bigint_impl {
...
+            #[inline]
+            fn into(self) -> BigUint {
+                use num_traits::Zero;
+
+                let mut sum = BigUint::zero();
+
+                for i in (0..$num_limbs).rev() {
+                    sum <<= 64;
+                    sum += self.0[i];
+                }
```
Comment (daira) [inline ff/src/biginteger/macros.rs:393]: This also seems to have quadratic cost in the number of limbs.

File: ff/src/biginteger/macros.rs (hunk 1 above)
Comment (weikengchen) [reply ff/src/biginteger/macros.rs:374]: You are right. Maybe I should use a better way: BigUint has `from_bytes_le` and `to_bytes_le`, from which I may have a better construction.

**Review by daira — APPROVED** [review]
On the topic of how explicit conversions should be, they absolutely need to be explicit between different fields, and between integers and fields, and from an integer to a potentially shorter integer type (i.e. where the value can change).

**Comment by weikengchen** [comment]
Two updates: (1) I changed the algorithm, which tries to use the to/from_bytes_le as much as possible, though sometimes I need to make do with the second-best (2) The two traits are implemented for `PrimeField` as well.

File: ff/src/biginteger/macros.rs [hunk 3]
```diff
@@ -352,5 +352,36 @@ macro_rules! bigint_impl {
...
+                        let mut chunk_padded = [0u8; 8];
+                        for j in 0..chunk.len() {
+                            chunk_padded[j] = chunk[j];
+                        }
+                        limbs[i] = u64::from_le_bytes(chunk_padded)
+                    });
+
+                Self(limbs)
+            }
+        }
```
Comment (Pratyush) [inline ff/src/biginteger/macros.rs:378]: Maybe we change this to `TryFrom`? In case the `BigUInt is bigger?

File: ff/src/fields/macros.rs [hunk 4]
```diff
@@ -717,5 +717,19 @@ macro_rules! impl_Fp {
                 self.0.zeroize();
             }
         }
+
+        impl<P: $FpParameters> From<num_bigint::BigUint> for $Fp<P> {
+            #[inline]
+            fn from(val: num_bigint::BigUint) -> $Fp<P> {
+                $Fp::<P>::from_repr(<$BigIntegerType>::from(val)).unwrap()
```
Comment (Pratyush) [inline ff/src/fields/macros.rs:724]: For this, let's use `from_bytes_mod_order`, or convert `P::MODULUS` to `BigUint` and then use the modulo operator there.

File: ff/src/biginteger/macros.rs (hunk 3 above)
Comment (weikengchen) [reply ff/src/biginteger/macros.rs]: Done.

File: ff/src/fields/macros.rs (hunk 4 above)
Comment (weikengchen) [reply ff/src/fields/macros.rs]: Done.

File: ff/src/biginteger/macros.rs (hunk 3 above)
Comment (weikengchen) [reply ff/src/biginteger/macros.rs]: (more fixings needed. on the way)

**Comment by weikengchen** [comment]
Done the changes.

File: ff/src/biginteger/macros.rs [hunk 5]
```diff
@@ -352,5 +352,45 @@ macro_rules! bigint_impl {
...
+
+                    bytes
+                        .chunks(8)
+                        .into_iter()
+                        .enumerate()
+                        .for_each(|(i, chunk)| {
+                            let mut chunk_padded = [0u8; 8];
+                            for j in 0..chunk.len() {
+                                chunk_padded[j] = chunk[j];
+                            }
```
Comment (daira) [inline ff/src/biginteger/macros.rs:380]:
```suggestion
                            chunk_padded[..chunk.len()].copy_from_slice(chunk);
```

**Review by daira — APPROVED** [review]
Untested ACK

**Comment by weikengchen** [comment]
Done.

**Comment by weikengchen** [comment]
All checks have passed, and the code is ready for review.

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,12 +12,13 @@
 - [\#261](https://github.com/arkworks-rs/algebra/pull/261) (ark-ff) Add support for 448-bit integers and fields.
 - [\#263](https://github.com/arkworks-rs/algebra/pull/263) (ark-ff) Add `From<iXXX>` implementations to fields.
 - [\#265](https://github.com/arkworks-rs/algebra/pull/265) (ark-serialize) Add hashing as an extension trait of `CanonicalSerialize`.
+- [\#280](https://github.com/arkworks-rs/algebra/pull/280) (ark-ff) Add `Into<BigUint>` and `From<BigUint>` implementations to `BigInteger` and `PrimeField`.
```
Comment (daira) [inline CHANGELOG.md:15]: This is outdated.

Comment (weikengchen) [reply CHANGELOG.md:15]: Good catch. Let me rename the PR as well.

**Review by daira — APPROVED** [review]
Untested ACK

**Comment by weikengchen** [comment]
Ready for review. 

The linter test failure is related to a separate reason: #283

**Review by Pratyush — APPROVED** [review]

**Pull request merged**